efi/fw_load_handler.go

```diff
@@ -20,9 +20,251 @@
 package efi
 
 import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	efi "github.com/canonical/go-efilib"
+	"github.com/canonical/go-tpm2"
 	"github.com/canonical/tcglog-parser"
+	"golang.org/x/xerrors"
+)
+
+const (
+	sbStateName = "SecureBoot" // Unicode variable name for the EFI secure boot configuration (enabled/disabled)
 )
 
+// fwLoadHandler is an implementation of imageLoadHandler that measures firmware
+// events (pre-OS events and events related to loading of OS components).
+type fwLoadHandler struct {
+	log *tcglog.Log
+}
+
 var newFwLoadHandler = func(log *tcglog.Log) imageLoadHandler {
-	panic("not implemented")
+	return &fwLoadHandler{log: log}
+}
+
+func (h *fwLoadHandler) measureSignatureDb(ctx pcrBranchContext, name efi.VariableDescriptor) ([]byte, error) {
+	db, _, err := ctx.Vars().ReadVar(name.Name, name.GUID)
+	if err != nil && err != efi.ErrVarNotExist {
+		return nil, xerrors.Errorf("cannot read current variable: %w", err)
+	}
+
+	ctx.MeasureVariable(secureBootPCR, name.GUID, name.Name, db)
+	return db, nil
+}
+
+func (h *fwLoadHandler) measureAuthorizedSignatureDb(ctx pcrBranchContext) error {
+	data, err := h.measureSignatureDb(ctx, Db)
+	if err != nil {
+		return err
+	}
+
+	db, err := efi.ReadSignatureDatabase(bytes.NewReader(data))
+	if err != nil {
+		return xerrors.Errorf("cannot decode signatures: %w", err)
+	}
+	ctx.FwContext().Db = &secureBootDB{Name: Db, Contents: db}
```

Comment:
would this kind of code be clearer as some kind of method on fwContext? also I notice that no method of fwContext actually uses Db, will the Db from the fwContext be used by later stages beyond the fw load handling ?

Reply:
`fwContext.Db` is used in this PR by `fwImageLoadMeasurer.measureVerification`, but it will also be used by the next PR as well.

Reply:
I'll try to keep an eye on Db usage in the next PRs, to see if we can have the fwContext have a more structured surface